### PR TITLE
Improve typing hints for Engine and helper methods

### LINKED ISSUE #2356: Improve typing hints for Engine and helper methods
## 🚀 Feature

1) Let's improve typing hints for process_function arg:
- https://github.com/pytorch/ignite/blob/8a1de70e2152936ffcb051f3bd618138130626ec/ignite/engine/engine.py#L122
We have to provide the signature: [Engine, Any] -> Any in mypy recognizable way

Also here: https://github.com/pytorch/ignite/blob/8a1de70e2152936ffcb051f3bd618138130626ec/ignite/engine/deterministic.py#L179

2) Let's improve typing hints for `prepare_batch` and `output_transform` for `create_supervised_trainer` and `create_supervised_evaluator`.

https://github.com/pytorch/ignite/blob/8a1de70e2152936ffcb051f3bd618138130626ec/ignite/engine/__init__.py#L52-L53

3) Provide example of usage of `prepare_batch` and `output_transform` in the docstring: https://pytorch.org/ignite/generated/ignite.engine.create_supervised_trainer.html

cc @Priyansi

### PR DESCRIPTION
**Pull request by louis-she**
Fixes #2356

Description:

Check list:

- [ ] New tests are added (if a new feature is added)
- [x] New doc strings: description and/or example code are in RST format
- [x] Documentation is updated (if required)

### CODE REVIEW

**Comment by louis-she** [comment]
https://github.com/pytorch/ignite/issues/2356

File: ignite/engine/deterministic.py [hunk 1]
```diff
@@ -176,7 +176,7 @@ class DeterministicEngine(Engine):
             in each iteration, and returns data to be stored in the engine's state.
     """
 
-    def __init__(self, process_function: Callable):
+    def __init__(self, process_function: Callable[["DeterministicEngine", Any], Any]):
```
Comment (vfdev-5) [inline ignite/engine/deterministic.py:179]: Looks like mypy is failing here, let's make the signature as `Callable[[Engine, Any], Any])`

**Comment by louis-she** [comment]
@vfdev-5 CI has passed

**Review by vfdev-5 — APPROVED** [review]
LGTM, thanks for the PR @louis-she !

**Pull request merged**